lisa/test_runner/tests/test_lisarunner.py

```diff
@@ -0,0 +1,411 @@
+        # every test case need a new environment, so must to create
+        for test_result in test_results:
+            test_result.runtime_data.use_new_environment = True
+        runner._merge_test_requirements(
+            test_results=test_results,
+            existing_environments=envs,
+            platform_type=constants.PLATFORM_MOCK,
+        )
+        self.assertListEqual(
+            ["runbook_0", "req_1", "req_2", "req_3"], [x for x in envs],
+        )
+        self.assertListEqual(
+            [TestStatus.NOTRUN, TestStatus.NOTRUN, TestStatus.NOTRUN],
+            [x.status for x in test_results],
+        )
+
+    def test_merge_req_not_allow_create(self) -> None:
+        env_runbook = generate_env_runbook(is_single_env=False)
+        env_runbook.allow_create = False
+        envs = load_environments(env_runbook)
+        self.assertListEqual(
+            [], [x for x in envs],
+        )
+        runner = generate_lisarunner(None)
+        test_results = generate_cases_result()
+        runner._merge_test_requirements(
+            test_results=test_results,
+            existing_environments=envs,
+            platform_type=constants.PLATFORM_MOCK,
+        )
+        self.assertListEqual(
+            [], [x for x in envs],
+        )
+        self.assertListEqual(
+            [TestStatus.SKIPPED, TestStatus.SKIPPED, TestStatus.SKIPPED],
+            [x.status for x in test_results],
+        )
+        not_allow_new_message = (
+            "not found fit environment, and not allow to create new environment"
+        )
+        self.assertListEqual(
+            [not_allow_new_message, not_allow_new_message, not_allow_new_message],
+            [x.message for x in test_results],
+        )
+
+    def test_merge_req_platform_type_checked(self) -> None:
+        # checked is enough. More covered in search space
+        env_runbook = generate_env_runbook(is_single_env=False)
+        envs = load_environments(env_runbook)
+        self.assertListEqual(
+            [], [x for x in envs],
+        )
+        runner = generate_lisarunner(None)
+        test_results = generate_cases_result()
+        for test_result in test_results:
+            metadata = test_result.runtime_data.metadata
+            metadata.requirement = simple_requirement(
+                supported_platform_type=["notexists"]
+            )
+        runner._merge_test_requirements(
+            test_results=test_results,
+            existing_environments=envs,
+            platform_type=constants.PLATFORM_MOCK,
+        )
+        self.assertListEqual(
+            [TestStatus.SKIPPED, TestStatus.SKIPPED, TestStatus.SKIPPED],
+            [x.status for x in test_results],
+        )
+        platform_unsupported = "capability cannot support some of requirement"
+        self.assertListEqual(
+            [platform_unsupported, platform_unsupported, platform_unsupported],
+            [x.message[0:45] for x in test_results],
```

Comment:
Might want to use `startswith()` instead of hardcoding a splice with a length.

Reply:
assertListEqual can return compared strings in error message, if there is failure. string methods like startswith will hide information, so that it have to debug to see detail messages.

Reply:
I believe Andy suggested something like:

```suggestion
            [platform_unsupported in x.message for x in test_results],
```

(In this case, `startswith()` isn't the right approach, if the message can appear anywhere in the string, but you can use the `in` operator.  Good idea to avoid hardcoded lengths and let the interpreter sort it out.)

Reply:
I don't it can appear anywhere in the string, right now we're splicing from the 0th index to 45. So the first 45 characters, right? I could be misreading it.

Reply:
Oh, yeah, you're right.
The suggestion would be:

```python
            [x.message.startswith(platform_unsupported) for x in test_results],
```

Then.

Reply:
Changed like this. The exepcted and actual messages will pass into assertListEqual. The assertListEqual compares length and content one by one. If anything mismatch, it prints both expected and actual strings. That's more details than False output by startswith or in.

```python
        # compare it's begin with
        actual_messages = [
            test_results[index].message[0 : len(expected)]
            for index, expected in enumerate(expected_message)
        ]
        self.assertListEqual(
            expected_message, actual_messages,
        )
```